### PR TITLE
Fixed #29697 -- Changed Query.trim_start for avoid removing the aux tables

### PR DESCRIPTION
**Pull request by cansarigol**
Ticket: [#29697](https://code.djangoproject.com/ticket/29697)

### CODE REVIEW

File: django/db/models/sql/query.py [hunk 1]
```diff
@@ -2076,13 +2076,22 @@ def trim_start(self, names_with_path):
             t for t in self.alias_map
             if t in self._lookup_joins or t == self.base_table
         ]
+        # from_opts_tables for manage to define unref_alias table.
+        # before unref, make sure the table wasn't added
+        # in another from_opts as reference.
+        from_opts_tables = [
+            path.from_opts.db_table for path in all_paths
+            if path.from_opts.db_table != self.base_table
+        ]
```
Comment (charettes) [inline django/db/models/sql/query.py]: Use a set since it's only used for containment checks.

File: tests/queries/tests.py [hunk 2]
```diff
@@ -3918,3 +3920,31 @@ def test_ticket_23622(self):
...
+        cls.u1 = Ticket29697User.objects.create(company=cls.c1, is_superuser=True, id=4)
+        cls.u2 = Ticket29697User.objects.create(company=cls.c2, is_superuser=True)
+        cls.u3 = Ticket29697User.objects.create(company=cls.c3, is_superuser=True)
+        cls.u4 = Ticket29697User.objects.create(company=cls.c4, is_superuser=True)
+
+    def test_ticket_29697(self):
+        status = Ticket29697Status.objects.filter(
+            Q(company__employees__is_superuser=True) |
+            ~Q(company__employees__permissions__contains=Concat(Value('records.'), 'key', Value('.denied'))),
+            company__employees__pk=4).order_by('id')
```
Comment (charettes) [inline tests/queries/tests.py]: I'd be great if you could make the test case more concise somehow. I doubt all of these criterias are required to hit the issue and they add a lot of noise which makes future debugging harder.

**Comment by cansarigol** [comment]
I tried to apply your recommendations. Could you check again?

File: django/db/models/sql/query.py [hunk 3]
```diff
@@ -2052,6 +2052,24 @@ def extra_select(self):
...
+        """
+        alias_joined_list = set([k for k, v in self.alias_refcount.items() if v > 0])
+        from_to_tables = set([
+            path.from_opts.db_table for path in all_paths
+            if path.from_opts.db_table in [
+                path.to_opts.db_table for path in all_paths
+                if path.from_opts.db_table in alias_joined_list and
+                path.from_opts.db_table != path.to_opts.db_table
+            ]
+        ])
```
Comment (charettes) [inline django/db/models/sql/query.py]: Use set literals, `{...}` here and above.

File: django/db/models/sql/query.py [hunk 4]
```diff
@@ -2052,6 +2052,24 @@ def extra_select(self):
...
             return self.extra
 
+    def is_alias_available_join(self, all_paths, alias):
+        """
+        Check alias for whether both from and to in all_paths.
+        If so shouldn't be removed
+        alias_refcount might be changed hence need alias_joined_list
+        to check available from join
+        """
+        alias_joined_list = set([k for k, v in self.alias_refcount.items() if v > 0])
```
Comment (charettes) [inline django/db/models/sql/query.py]: Might want to rename to something else since it's not a list?

File: django/db/models/sql/query.py [hunk 5]
```diff
@@ -2076,13 +2094,17 @@ def trim_start(self, names_with_path):
             t for t in self.alias_map
             if t in self._lookup_joins or t == self.base_table
         ]
+        # from_opts_tables for manage to define unref_alias table.
+        # before unref, make sure the table wasn't added
+        # in another from_opts as reference.
```
Comment (charettes) [inline django/db/models/sql/query.py]: I guess this comment can be removed now?

File: tests/queries/tests.py [hunk 6]
```diff
@@ -1199,6 +1199,16 @@ def test_common_mixed_case_foreign_keys(self):
         )
         self.assertTrue(qs.first())
 
+    def test_ticket_29697_complex_query_crashes(self):
```
Comment (charettes) [inline tests/queries/tests.py]: `test_excluded_intermediary_m2m_table_joined`?

**Comment by cansarigol** [comment]
@charettes thanks a lot. sorry about these shortcomings and thank again for your help.

File: django/db/models/sql/query.py [hunk 7]
```diff
@@ -2082,7 +2100,8 @@ def trim_start(self, names_with_path):
             if self.alias_map[lookup_tables[trimmed_paths + 1]].join_type == LOUTER:
                 contains_louter = True
             alias = lookup_tables[trimmed_paths]
-            self.unref_alias(alias)
+            if self.is_alias_available_join(all_paths, alias):
```
[the remark this answers is not in the document]
Comment (cansarigol) [reply django/db/models/sql/query.py]: I'm not sure whether is_alias_available_join is useful. the tables I am interested at will not be there. so I removed it

File: django/db/models/sql/query.py [hunk 8]
```diff
@@ -2052,6 +2052,24 @@ def extra_select(self):
         else:
             return self.extra
 
+    def is_alias_available_join(self, all_paths, alias):
+        """
+        Check alias for whether both from and to in all_paths.
+        If so shouldn't be removed
+        alias_refcount might be changed hence need joined_alias
+        to check available from join
+        """
```
Comment (charettes) [inline django/db/models/sql/query.py]: This docstring needs rewording. It should focus on what the function is doing and not on what the called should do.

File: django/db/models/sql/query.py [hunk 9]
```diff
@@ -2052,6 +2052,24 @@ def extra_select(self):
...
+        joined_alias = {k for k, v in self.alias_refcount.items() if v > 0}
+        from_to_tables = {
+            path.from_opts.db_table for path in all_paths
+            if path.from_opts.db_table in {
+                path.to_opts.db_table for path in all_paths
+                if path.from_opts.db_table in joined_alias and
+                path.from_opts.db_table != path.to_opts.db_table
+            }
+        }
+        return alias not in from_to_tables
```
Comment (charettes) [inline django/db/models/sql/query.py]: This seems overly complicated and probably also wrong with tables that are JOIN'ed more than once as the following JOINs will have aliases that are not `db_table`.

Also you probably don't need to build the `from_to_tables` set as it's really inneficent. You should use a loop and exit the function as soon as a matching alias is found.

**Comment by cansarigol** [comment]
thanks for your reviews. @charettes you're right about alias, I think so too but I couldn't find a better solution, it exceeded me. I changed is_alias_available_join

**Comment by charettes** [comment]
> you're right about alias, I think so too but I couldn't find a better solution, it exceeded me.

This part of the ORM is definitely tricky, it's normal that you are feeling overwhelmed. I still think we should find a way to test the multi-alias case and address it before merging this change though.

**Comment by cansarigol** [comment]
hi @charettes, pushed a commit about maintaining is available check func before unref_alias and test for multi-alias queries.

**Comment by felixxm** [comment]
The added test passes without changes in `Query`. Bug was probably fixed by Simon refactorization or test doesn't cover this issue. I believe in Simon :smile: , so I will check which commit has fixed this.

**Comment by felixxm** [comment]
Bug was fixed by f19a4945e1191e1696f1ad8e6cdc6f939c702728.

Tests updated in #11142.

**Comment by cansarigol** [comment]
that's a great solution. @charettes :+1:

**Pull request closed without merging**